### ISSUE TITLE
Exported PSK files do not contain smoothing groups

### ISSUE THREAD

**Issue by cmbasnett** [gildor2/UEViewer#235]
I have extracted PSK meshes out of some UT2004 packages that had proper smoothing groups at the time of import.

The umodel export of those meshes have the smoothing group information stripped out of them.

I'm not sure if this is a umodel problem or not, but if it is possible to get the smoothing group information out, our team would love this.

**Comment by Blenux** [comment]
From what I can remember, the PSK file format doesn't support smoothing groups, if you want them,  have to export with gLTF instead.

**Comment by gildor2** [comment]
PSK has smoothing groups. Unreal doesn't have them, so they're "exported" zeroed by UModel.
Unreal has normals, which are exported to gltf (as Blenux said).

**Comment by cmbasnett** [comment]
For some reason I was under the impression that the PSK data just sits raw inside packages after being imported. Please correct me if I'm wrong about that.

If that's not the case, the smoothing groups could be inferred (regenerated) based on the vertex normals. Presumably this isn't being done because it's a hassle.

**Comment by gildor2** [comment]
> For some reason I was under the impression that the PSK data just sits raw inside packages after being imported. Please correct me if I'm wrong about that.

It is not true for any UE1-UE4 engine. Everything is converted into internal data. Smoothing groups are used to generate normals, then data is removed.

> If that's not the case, the smoothing groups could be inferred (regenerated) based on the vertex normals. Presumably this isn't being done because it's a hassle.

Yes, it is very hard to rebuild smoothing groups based on normals. I saw some open-source applications doing that, but looking at implementation I always see pitfalls, it's too easy to make a mesh which will break this regeneration. At the same time, I don't want to waste my own time trying to do that myself.

**Comment by cmbasnett** [comment]
Fair enough. Thanks for the insight! Perhaps it would be better if GLTF was the default export format instead of PSK/PSA, since GLTF doesn't lose the shading information.

**Comment by gildor2** [comment]
You may change defaults in UModel's settings.
